### ISSUE TITLE
Android 12 - IllegalAccessException: Tried to access visual service WindowManager from a non-visual Context

### ISSUE THREAD

**Issue by ribafish** [BranchMetrics/android-branch-deep-linking-attribution#977]
Hi,

When running on API 31 device, we're getting the following log in logcat:

```
E/ContextImpl: Tried to access visual service WindowManager from a non-visual Context:com.globalid.App@9687de7 WindowManager should be accessed from Activity or other visual Context. Use an Activity or a Context created with Context#createWindowContext(int, Bundle), which are adjusted to the configuration and visual bounds of an area on screen.
    java.lang.IllegalAccessException: Tried to access visual service WindowManager from a non-visual Context:com.example.App@9687de7
        at android.app.ContextImpl.getSystemService(ContextImpl.java:2059)
        at android.content.ContextWrapper.getSystemService(ContextWrapper.java:857)
        at io.branch.referral.SystemObserver.getScreenDisplay(SystemObserver.java:375)
        at io.branch.referral.DeviceInfo.updateRequestWithV1Params(DeviceInfo.java:66)
        at io.branch.referral.ServerRequest.setPost(ServerRequest.java:193)
        at io.branch.referral.ServerRequestRegisterClose.<init>(ServerRequestRegisterClose.java:35)
        at io.branch.referral.Branch.executeClose(Branch.java:873)
        at io.branch.referral.Branch.closeSessionInternal(Branch.java:830)
        at io.branch.referral.BranchActivityLifecycleObserver.onActivityStopped(BranchActivityLifecycleObserver.java:105)
        at android.app.Application.dispatchActivityStopped(Application.java:476)
        at android.app.Activity.dispatchActivityStopped(Activity.java:1463)
        at android.app.Activity.onStop(Activity.java:2622)
        at androidx.fragment.app.FragmentActivity.onStop(FragmentActivity.java:496)
        at androidx.appcompat.app.AppCompatActivity.onStop(AppCompatActivity.java:252)
        at android.app.Instrumentation.callActivityOnStop(Instrumentation.java:1496)
        at android.app.Activity.performStop(Activity.java:8281)
        at android.app.ActivityThread.callActivityOnStop(ActivityThread.java:5148)
        at android.app.ActivityThread.performStopActivityInner(ActivityThread.java:5128)
        at android.app.ActivityThread.handleStopActivity(ActivityThread.java:5193)
        at android.app.servertransaction.StopActivityItem.execute(StopActivityItem.java:43)
        at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

This is on branch 5.0.15 and 5.1.2. It only happens when running on device with Android 12. Our app is targeting API 31. 

Same thing happens on activity start as well.

**Comment by gdeluna-branch** [comment]
Hi ribafish,

Thanks for reporting this. 

Can you:
describe any repro steps e.g foregrounding, backgrounding/closing the app and how consistent it is.
share the device model.
also attach the activity start stack trace, just to confirm the same error path in case there need to be fixes anywhere else.

Thanks

**Comment by ribafish** [comment]
Sure thing. We've noticed this when opening the app `MainActivity.onStart()`, where we call (in our `BranchManager`, as you'll see in the trace)
```
Branch.sessionBuilder(activity)
                .withCallback(this::handleDeepLink)
                .init()
```

And the error reported I pasted above is when backgrounding the app, which calls `MainActivity.onStop()` and also triggers your `BranchActivityLifecycleObserver.onActivityStopped`.


The log for app start error:
```
E/ContextImpl: Tried to access visual service WindowManager from a non-visual Context:com.example.App@a408f00 WindowManager should be accessed from Activity or other visual Context. Use an Activity or a Context created with Context#createWindowContext(int, Bundle), which are adjusted to the configuration and visual bounds of an area on screen.
    java.lang.IllegalAccessException: Tried to access visual service WindowManager from a non-visual Context:com.globalid.App@a408f00
        at android.app.ContextImpl.getSystemService(ContextImpl.java:2059)
        at android.content.ContextWrapper.getSystemService(ContextWrapper.java:857)
        at io.branch.referral.SystemObserver.getScreenDisplay(SystemObserver.java:375)
        at io.branch.referral.DeviceInfo.updateRequestWithV1Params(DeviceInfo.java:66)
        at io.branch.referral.ServerRequest.setPost(ServerRequest.java:192)
        at io.branch.referral.ServerRequestInitSession.setPost(ServerRequestInitSession.java:48)
        at io.branch.referral.ServerRequestRegisterOpen.<init>(ServerRequestRegisterOpen.java:30)
        at io.branch.referral.Branch.getInstallOrOpenRequest(Branch.java:1951)
        at io.branch.referral.Branch$InitSessionBuilder.init(Branch.java:3117)
        at net.example.branch.BranchManager.onActivityStarted(BranchManager.kt:72)
        at com.example.ui.main.MainActivity.onStart(MainActivity.kt:60)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1455)
        at android.app.Activity.performStart(Activity.java:8076)
        at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3660)
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:221)
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

**Comment by gdeluna-branch** [comment]
Perfect, thanks for the details. I'll provide an update next week.

**Comment by gdeluna-branch** [comment]
Update: I was able to reproduce exactly as described when I enabled StrictMode on API 31. I have a fix here https://github.com/BranchMetrics/android-branch-deep-linking-attribution/pull/979
awaiting review from team.

Thanks

**Comment by gdeluna-branch** [comment]
Released as part of 5.1.3
https://github.com/BranchMetrics/android-branch-deep-linking-attribution/releases/tag/5.1.3
https://search.maven.org/artifact/io.branch.sdk.android/library/5.1.3/aar

Please reopen if this continues to repro.

Thanks!
